### PR TITLE
Add a personal MCP gateway with scoped OAuth and tool discovery

### PR DESCRIPTION
**Pull request by basnijholt**
External MCP clients can use a personal agent’s assigned tools through an opt-in `/mcp` endpoint. Authorization reuses the Connections portal’s signed browser identity and personal provider connections, then issues a separate OAuth grant for each client connection. Users connect only the services they need.

- Expose only `search_tools`, `get_tool`, and `invoke_tool`, with bounded discovery, schemas, arguments, and results. Missing credentials or unrelated provider outages do not block other integrations.
- Issue durable resource-bound OAuth grants with PKCE S256, explicit consent, rotating refresh, replay detection, and current access checks. Bound onboarding, token issuance and retained state.
- Add connected-client management in the personal portal: client address, last tool use, inactivity and absolute expiry, pagination, Disconnect, and Disconnect all. Revocation preserves provider connections and prevents already-open consent from restoring removed grants.
- Add a restricted SCIM User lifecycle endpoint with a dedicated provisioning secret. Disabling, renaming or deleting a provisioned account atomically revokes its grants and pending consent. Re-enabling requires fresh approval.
- Keep access tokens at 15 minutes. With provisioning configured, grants default to 30 days idle and 180 days absolute; successful refresh extends idle expiry. Without provisioning, the 30-day absolute limit remains. Compact refresh history so ordinary six-month use fits the existing requester storage budget.
- Reuse personal tool filters, hooks and worker routing. Revalidate current configuration at the provider entrypoint after preparation, hooks and thread scheduling, so a permission change blocks a stale prepared invocation. Add stateless Streamable HTTP, deadlines, grant-scoped cancellation, and configurable active-call limits per grant, requester and process (defaults remain 16/32/128). Harden requester namespaces, route authorization, JWT/protocol input bounds, and SDK error privacy.

**Upgrade impact:** Existing requester-scoped workers and connections require reprovisioning or reconnection because the new exact-identity namespace cannot adopt ambiguous legacy ownership. Existing OAuth grants keep their original absolute deadlines. Enabling managed-account mode requires older unbound client grants to receive fresh approval once.

Validation: full `uv run pytest -n 4 --no-cov` passed **16,171 tests**, with 22 skipped, including the HTTP boundary cases and 24 sync/async configuration-publication regressions. The 152 focused gateway tests also cover configurable limits above and below defaults, environment wiring across user/client grants, invalid settings, and capacity retained during cancellation cleanup. The frontend suite passed **558 tests**, with one skipped. Full `uv run pre-commit run --all-files`, type/import-boundary checks and both frontend builds passed. Independent task and whole-branch reviews have no outstanding findings.

Actual local HTTPS acceptance passed **18 groups** using Chromium, MCP SDK 1.30, two isolated signed identities, controlled provider APIs and a 125-function MCP subprocess. It covered optional service linking, bounded discovery, private credential isolation, refresh, provider outages, client disconnect controls, account disable across restart, and fresh approval after reactivation. Source/build hashes stayed unchanged and all owned services stopped. These fixtures do not certify live identity-provider or SaaS enrollment.

The gateway is disabled by default and supports one API process. Browser logout does not revoke independent client grants. SCIM offboarding takes effect when the update is received and committed; delivery delay and already-dispatched actions remain deployment concerns. Live identity-provider enrollment must validate the restricted User profile, exact verified-email matching, and disabled group management.

Per-user credential routing does not provide an OS sandbox. Execution tools need isolated workers for mutually untrusted users; native integrations share the API process. Stuck work may require supervisor termination. Logical storage budgets also require physical volume limits. Tools requiring approval or a live chat context are unavailable.

See `docs/deployment/personal-mcp-gateway.md` for setup, proxy routes, lifecycle, provisioning and compatibility.

## Summary by CodeRabbit

* **New Features**
  * Added an optional Personal MCP Gateway for discovering, viewing schemas, and invoking assigned tools.
  * Added OAuth/PKCE authorization, client management, configurable lifetimes and concurrency limits, usage quotas, and SCIM 2.0 account provisioning.
  * Added Connections page controls for viewing and disconnecting authorized MCP clients.

* **Bug Fixes**
  * Improved request validation, cancellation handling, storage cleanup, and configuration-change protection.
  * Improved trusted-upstream authentication security and requester-scoped worker isolation.

* **Documentation**
  * Added deployment, authentication, MCP, sandbox, and worker-runtime guidance.

### CODE REVIEW

File: src/mindroom/mcp_gateway/oauth.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+        def save(connection: sqlite3.Connection) -> None:
+            existing = connection.execute(
+                "SELECT metadata FROM clients WHERE client_id = ?",
+                (normalized.client_id,),
+            ).fetchone()
+            if existing and existing["metadata"] != normalized.model_dump_json():
+                raise RegistrationError("invalid_client_metadata", "Client is already registered")  # noqa: EM101
+            connection.execute(
+                "INSERT OR IGNORE INTO clients VALUES (?, ?)",
+                (normalized.client_id, normalized.model_dump_json()),
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/mcp_gateway/oauth.py]: Fixed in eda893df6. Public onboarding now has configurable rate and byte budgets, fixed abandoned-registration retention, and transactional cleanup. Active grants stay usable at capacity; live-family access/refresh bindings preserve revocation and replay detection. Added concurrent admission, UTF-8 accounting, disk reuse, migration, and refresh/revocation regressions. Full suite: 15,950 passed, 22 skipped. Final real HTTPS/browser/SDK acceptance: all 13 groups passed.

File: src/mindroom/mcp_gateway/store.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                CREATE TABLE IF NOT EXISTS grants (
+                    grant_id TEXT PRIMARY KEY, payload TEXT NOT NULL, expires_at REAL NOT NULL,
+                    revoked INTEGER NOT NULL DEFAULT 0
+                );
+                CREATE TABLE IF NOT EXISTS capabilities (
+                    token_hash TEXT PRIMARY KEY, kind TEXT NOT NULL, grant_id TEXT NOT NULL,
+                    payload TEXT NOT NULL, expires_at REAL NOT NULL, consumed INTEGER NOT NULL DEFAULT 0,
+                    FOREIGN KEY (grant_id) REFERENCES grants(grant_id)
+                );
+                CREATE INDEX IF NOT EXISTS capabilities_grant ON capabilities(grant_id);
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/mcp_gateway/store.py:70]: Fixed in eda893df6. Write transactions and client-registration lookups collect expired onboarding state and inactive families. Anonymous onboarding also has an aggregate byte budget and admission rate limit. Live-family access/refresh bindings are retained: deleting a binding immediately after refresh could otherwise break revocation that had already loaded that token. Expiry, replay, concurrent revocation, repeated disk reuse, and legacy migration are covered by regressions.

File: src/mindroom/mcp_gateway/store.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    def _connect(self) -> sqlite3.Connection:
+        connection = sqlite3.connect(self.path, timeout=10, isolation_level=None)
+        connection.row_factory = sqlite3.Row
+        connection.execute("PRAGMA foreign_keys = ON")
+        return connection
+
+    def _transaction(self, operation: Callable[[sqlite3.Connection], _T]) -> _T:
+        connection = self._connect()
+        try:
+            connection.execute("BEGIN IMMEDIATE")
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/mcp_gateway/store.py]: Fixed the capability lookup contention in eda893df6 with query-only, deferred reads and a one-second busy timeout. Code, access-token, and refresh-token lookups use that path; exchanges and revocation still revalidate atomically under BEGIN IMMEDIATE. The regression holds an unrelated writer reservation while all three lookups complete, then verifies stale models cannot bypass committed revocation. get_client remains on the write path because it now enforces registration retention. We retained rollback journaling, consistent with configure_state_engine_pragmas in src/mindroom/agent_storage.py and the existing OAuth credential store, to preserve filesystem and single-file-backup compatibility. SQLite documents the relevant [WAL filesystem restriction](https://sqlite.org/wal.html).

File: src/mindroom/api/mcp_gateway.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    if request.method == "OPTIONS":
+        return Response(
+            status_code=204,
+            headers={
+                **_MACHINE_HEADERS,
+                "Access-Control-Allow-Methods": "POST, OPTIONS",
+                "Access-Control-Allow-Headers": "Content-Type, MCP-Protocol-Version",
+            },
+        )
+    if operation in {"register", "authorize"} and not runtime._allow_onboarding_request():
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/api/mcp_gateway.py]: Fixed in 2f8655773. Registration and authorization now share a per-source limit (default 10/minute) alongside the aggregate limit (60/minute). The bucket uses only the canonical ASGI client IP, including IPv4-mapped normalization; ports and raw forwarding headers cannot select a new bucket. Trusted proxy normalization remains with the ASGI server and is documented along with NAT/fallback sharing.

Only accepted requests occupy the bounded deque and source counters. Denied attempts neither consume aggregate capacity nor extend the window. HTTP regressions cover two sources in one runtime, spoofed headers, changed ports, aggregate saturation, unknown peers, exact expiry, and existing grant exchange/refresh/revocation.

All 24 gateway API tests, the full 15,955-test suite, full hooks, and the repeated 13-group local HTTPS/SDK/Chromium acceptance passed. Independent scoped review found no remaining issue.

**Comment by basnijholt** [comment]
@coderabbitai review

File: skills/mindroom-docs/references/page__deployment__personal-mcp-gateway__index.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+Keep that directory private and persistent across restarts.
+Opaque authorization codes, browser nonces, access tokens, and refresh tokens are stored as hashes; upstream provider credentials remain separate.
+Run one API process for this gateway: active-call limits and cancellation ownership are process-local.
+Multiple independently routed replicas are not supported by this implementation.
+
+Public registration and authorization starts share an aggregate admission limit of 60 requests per minute per API process and a per-source limit of 10 requests per minute.
+Set the positive integers `MINDROOM_MCP_GATEWAY_ONBOARDING_RATE_LIMIT` and `MINDROOM_MCP_GATEWAY_ONBOARDING_SOURCE_RATE_LIMIT` to adjust these limits.
+Keep the source limit below the aggregate limit to leave capacity for other sources.
+Excess requests receive HTTP 429 with `Retry-After`; token exchange, refresh, revocation, discovery, and existing MCP grants remain usable.
+Rejected requests do not consume another source's allowance or extend the rate-limit window.
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply skills/mindroom-docs/references/page__deployment__personal-mcp-gateway__index.md]: Fixed in c9d76a7d7. The source documentation now says rejected requests consume neither the per-source nor aggregate allowance and extend neither window. Both generated references were regenerated, and all repository pre-commit hooks passed. Runtime code is unchanged.

File: src/mindroom/mcp_gateway/server.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        """Run exactly once per owning API lifespan."""
+        async with self._manager.run():
+            try:
+                yield
+            finally:
+                self._closing = True
+                leases = tuple(self._active.values())
+                for lease in leases:
+                    lease.cancel()
+                await asyncio.gather(*(lease.wait() for lease in leases))
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/mcp_gateway/server.py:256]: Verified: a hung native call or cleanup can retain its slot and delay graceful shutdown indefinitely. The source documentation and PR description now state this explicitly, including the need for deployment-supervisor termination when a bounded process shutdown is required.

Hard termination needs a separate worker-process design. Native execution currently carries live requester identity, credentials, configuration, toolkit/hooks, and MCP lifecycle owners; existing worker routing does not cover that full lifecycle. A timeout around the drains cannot terminate the default-executor thread and risks tearing down resources while retained work still owns them. The capacity fix intentionally preserves that ownership.

An independent source review reached the same conclusion: no narrow in-process change can safely terminate arbitrary running native code. I am retaining this as a documented architectural limitation for this opt-in, single-process gateway, with hard provider isolation deferred. The HTTP regressions verify bounded admission, prompt cancellation responses, retention until actual completion, and rejection of new calls during shutdown. Bounded native-process shutdown remains unsupported.

File: src/mindroom/api/mcp_gateway.py [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+    if request.method == "OPTIONS":
+        return Response(
+            status_code=204,
+            headers={
+                **_MACHINE_HEADERS,
+                "Access-Control-Allow-Methods": "POST, OPTIONS",
+                "Access-Control-Allow-Headers": "Content-Type, MCP-Protocol-Version",
+            },
+        )
+    if operation in {"register", "authorize"} and not runtime._onboarding_limiter.allow(_onboarding_source(request)):
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/api/mcp_gateway.py:307]: This limiter bounds requests admitted to body parsing and OAuth handlers, including malformed attempts. Refunding failed validation would allow a source to repeat that work without consuming its admission budget. Requests already rejected with HTTP 429 are not recorded and do not extend either window.

The default per-source limit is 10 and aggregate limit is 60, so one source cannot exhaust the aggregate allowance; canonical source grouping and cross-source fairness are covered by the API tests. Multiple sources and callers sharing a NAT remain subject to the documented limits.

The documentation's phrase “rejected requests” is ambiguous. I am clarifying that it means rate-limit rejections and adding an explicit malformed-traffic fairness regression. Admission behavior should remain unchanged. Please reassess the proposed refund against this abuse-control boundary.

File: src/mindroom/mcp_gateway/toolkits.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    task = asyncio.create_task(cleanup, name="mcp-gateway-tool-cleanup")
+    retain_execution_task(task)
+    _CLEANUP_TASKS.add(task)
+
+    def finished(completed: asyncio.Task[None]) -> None:
+        _CLEANUP_TASKS.discard(completed)
+        if not completed.cancelled():
+            completed.exception()
+
+    task.add_done_callback(finished)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/mcp_gateway/toolkits.py:55]: Fixed in 72f151a26. Failed retained cleanup now logs `mcp_gateway_tool_cleanup_failed` with only the exception type; cancelled tasks remain silent. Task retention, removal, and draining are unchanged.

The new native-tool cancellation regression proves capacity remains occupied while close is pending, recovers after close fails, and emits exactly one warning without the provider detail from the exception message. All 139 focused tests, the full 15,967-test local suite, and all repository hooks passed.

**Comment by basnijholt** [comment]
@coderabbitai review

**Comment by basnijholt** [comment]
@coderabbitai review

Addressed the query-plan portability note in 45c094bf2. The test accepts both SQLite index annotations for the exact expected table/index pairs; full-history scans and other indexes remain rejected. This follow-up changes two test lines only. All 16,046 local tests and full repository hooks passed.

**Comment by basnijholt** [comment]
@coderabbitai review

File: src/mindroom/mcp_gateway/tools.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+
+async def _selected_operation[T](
+    context: PersonalAgentContext,
+    name: str,
+    manager: MCPServerManager | None,
+    operation: Callable[[Toolkit], Awaitable[T]],
+) -> T:
+    async def selected() -> T:
+        entry = await run_gateway_sync(_require_entry, context, name)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/mcp_gateway/tools.py:166]: Fixed in 5fbfcd0d3. The API supplies a live configuration check at the selected provider entrypoint, after toolkit preparation, awaited hooks, and sync thread scheduling. All 18 changed-configuration cases failed before the fix with provider execution, then passed; six equivalent-publication controls continue to succeed. The 24 HTTP regressions also verify cleanup and original identity preservation.

File: skills/mindroom-docs/references/page__deployment__personal-mcp-gateway__index.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+The gateway validates SDK request and notification envelopes before dispatch and returns fixed errors without logging malformed input or unknown tool names.
+Valid unsolicited response/error envelopes are acknowledged with an empty HTTP 202 and dropped: this stateless gateway never sends correlated client-result requests.
+If bidirectional requests or stateful sessions are added later, their client responses must instead reach the owning session.
+Client applications need support for this OAuth registration and discovery flow; a static bearer-only configuration screen cannot perform initial login.
+
+## Access and lifecycle
+
+Access tokens last up to 15 minutes.
+Refresh tokens rotate on use; replay revokes the grant family.
+A managed-account grant expires after 30 days without successful refresh or tool use, and at most 180 days after approval.
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply skills/mindroom-docs/references/page__deployment__personal-mcp-gateway__index.md]: Fixed in 5fbfcd0d3. Updated the source deployment guide to identify 30-day idle and 180-day absolute lifetimes as configurable defaults, then regenerated both reference files.

**Comment by basnijholt** [comment]
@coderabbitai review

Please review the latest follow-up: final native dispatch configuration revalidation and its HTTP regressions. The prior documentation finding and both JSDoc nitpicks are also addressed.

**Comment by basnijholt** [comment]
@coderabbitai review

Please review the cleanup follow-up: consolidated portal disconnect handling, runtime-owned SCIM account routing, and the supported-field set. Existing API behavior and lifecycle policy are unchanged.

File: src/mindroom/api/mcp_scim.py [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+    from mindroom.mcp_gateway.accounts import GatewayAccounts
+
+_BASE = "/mcp/scim/v2"
+_CORE = "urn:ietf:params:scim:schemas:core:2.0:"
+_MESSAGES = "urn:ietf:params:scim:api:messages:2.0:"
+_USER_SCHEMA = _CORE + "User"
+_HEADERS = {"Cache-Control": "private, no-store", "X-Content-Type-Options": "nosniff"}
+_MAX_BODY = 65_536
+_FILTER = re.compile(r'\s*(userName|emails\.value|id)\s+eq\s+("(?:[^"\\]|\\.)*")\s*', re.IGNORECASE)
+_FIELDS = {"username", "active", "displayname", "externalid", "name", "emails"}
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/api/mcp_scim.py:36]: Confirmed that these paths are unsupported, and that this predates the cleanup. `emails` is writable through PUT, pathless PATCH, and whole-attribute PATCH; the supported PATCH path set did not change in this commit.

This endpoint intentionally implements a restricted User lifecycle profile for compatible connectors. We are keeping email value-path parsing outside this behavior-preserving cleanup. The deployment guide now explicitly lists supported PATCH paths, the whole-`emails` update route, and the requirement to test actual connector update/deactivation payloads. Unsupported operations reject the entire PATCH atomically, so a connector combining an unsupported email path with deactivation must be adapted before offboarding can rely on it.

Verified this behavior against the real HTTP handlers and SQLite store. Independent review also classified it as an existing interoperability limitation, rather than a cleanup regression. Full email value-path support should be a separately scoped connector compatibility change with selection, no-match, primary-email, and rollback coverage.

**Comment by basnijholt** [comment]
@coderabbitai review
